pkg/placement/declarations.go

```diff
@@ -0,0 +1,387 @@
+	whatResolver WhatResolver,
+	whereResolver WhereResolver,
+	setBinder SetBinder,
+	workloadProjector WorkloadProjector,
+	placementProjector PlacementProjector,
+) {
+	whatResolver.AddConsumer(setBinder.AsWhatConsumer(), true)
+	whereResolver.AddConsumer(setBinder.AsWhereConsumer(), true)
+	workloadProjector.SetProvider(setBinder)
+	placementProjector.SetProvider(setBinder)
+}
+
+// ResolvedWhere identifies the set of SyncTargets that match a certain
+// EdgePlacement's "where" predicate.
+// Each `*edgeapi.SinglePlacementSlice` points to an immutable object.
+type ResolvedWhere []*edgeapi.SinglePlacementSlice
+
+// WorkloadParts identifies a workload prescription and provides
+// ephemeral details of how to access it.
+// A workload prescription is the things that match the "what" predicate
+// of an EdgePlacement.
+//
+// Every WorkloadParts that appears in the interfaces here is immutable.
+//
+// In the case of a Namespace object, this implies that
+// all the objects in that namespace are included.
+//
+// A workload may include objects of kinds that are built into
+// the edge cluster.  By built-in we mean that these kinds are both
+// already known (regardless of whether it is via being built into
+// the apiserver or added to it by either form of aggregation) and
+// not managed by edge workload management.
+// It is the user's responsibility to make the "what" predicate
+// match the corresponding CRD when the workload includes an object
+// of a kind that is not built into the edge cluster.
+type WorkloadParts map[WorkloadPartID]WorkloadPartDetails
+
+// WorkloadPartID identifies part of a workload.
+type WorkloadPartID struct {
+	APIGroup string
+
+	// Resource is the lowercase plural way of identifying the kind of object
+	Resource string
+
+	Name string
+}
+
+// WorkloadPartDetails provides additional details about how the WorkloadPart
+// is to be included.
+type WorkloadPartDetails struct {
+	// APIVersion is version (no group) that the source workspace prefers to serve.
+	APIVersion string
+
+	// IncludeNamespaceObject is only interesting for a Namespace part, and
+	// indicates whether to include the details of the Namespace object;
+	// the objects in the namespace are certainly included.
+	// For other parts, this field holds `false`.
+	IncludeNamespaceObject bool
+}
+
+type WorkloadPart struct {
+	WorkloadPartID
+	WorkloadPartDetails
+}
+
+// ProjectionMapProducer tells the consumers what to project,
+// organized into three levels.
+type ProjectionMapProducer DynamicMapProducer[ProjectionKey, *ProjectionPerCluster]
+
+// ProjectionKey identifies the topmost level of organization,
+// the combinatin of the destination and the API group and resource.
+type ProjectionKey struct {
+	metav1.GroupResource
+	Destination SinglePlacement
+}
+
+// ProjectionPerCluster is the second level of organization.
+// It identifies the API version to use currently and holds
+// the map provider that gets to the lowest level of organization.
```

Comment:
I found 'map provider' and 'map producer' are both used here and it seems that two terms refer to the same thing. We may want to be consistent for easier reading.

Reply:
In the latest push I tried to rationalize the terminology.